### PR TITLE
Update spec MetricNode for dbt x MetricFlow integration

### LINKED ISSUE #7500: [CT-2522] Redefine Metric node for the Manifest and update parsing
There are two new node types being added to the manifest as part of the MetricFlow integration work: SemanticModel nodes and Metric nodes. This ticket addresses the Metric nodes. In [dbt-semantic-interfaces](https://github.com/dbt-labs/dbt-semantic-interfaces) we're adding a protocol of a [Metric](https://github.com/dbt-labs/dbt-semantic-interfaces/issues/18), which will have the shared standard definition of a Metric. In dbt-core we need to compose this protocol with base node classes to define dbt-core's representation of a Metric. Additionally there are some additional fields we should expose as defined in #7456

### Blocked by:

- https://github.com/dbt-labs/dbt-core/issues/7470
- https://github.com/dbt-labs/dbt-semantic-interfaces/issues/18

### PR DESCRIPTION
**Pull request by QMalcolm**
resolves #7500 #7404

### Checklist

- [X] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [X] I have run this code in development and it appears to resolve the stated issue
- [X] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: core/dbt/contracts/graph/nodes.py [hunk 1]
```diff
@@ -1285,27 +1311,76 @@ def same_contents(self, old: Optional["Exposure"]) -> bool:
 # ====================================
 
 
+@dataclass
+class WhereFilter(dbtClassMixin, Replaceable):
```
Comment (peterallenwebb) [inline core/dbt/contracts/graph/nodes.py:1315]: Gerda recommended in my review that we not use Replaceable unless we need it, and I don't think we need it on these classes.

File: core/dbt/parser/schema_yaml_readers.py [hunk 2]
```diff
@@ -125,6 +144,102 @@ def __init__(self, schema_parser: SchemaParser, yaml: YamlBlock):
         self.schema_parser = schema_parser
         self.yaml = yaml
 
+    def _get_input_measure(
```
Comment (peterallenwebb) [inline core/dbt/parser/schema_yaml_readers.py:147]: These helper functions look good enough for a beta, but I think we should look at ways to make the code less repetitive and mechanical.

More specifically, if we are going to continue along these lines, it would be better to have a way to declare how the unparsed objects map to the parsed objects, rather than performing the conversion manually for these and every new object.

File: core/dbt/contracts/graph/manifest.py [hunk 3]
```diff
@@ -1319,14 +1319,7 @@ class WritableManifest(ArtifactMixin):
...
     def compatible_previous_versions(self):
-        return [
-            ("manifest", 4),
-            ("manifest", 5),
-            ("manifest", 6),
-            ("manifest", 7),
-            ("manifest", 8),
-            ("manifest", 9),
-        ]
+        return []
```
Comment (peterallenwebb) [inline core/dbt/contracts/graph/manifest.py:1322]: As discussed in slack, we'll probably want to add these back.

**Comment by QMalcolm** [comment]
I'd prefer to rebase off of main, but for the sanity of those in the midst of reviewing, I'm gonna merge main into this branch

File: core/dbt/contracts/graph/manifest.py [hunk 4]
```diff
@@ -985,6 +988,20 @@ def analysis_lookup(self) -> AnalysisLookup:
             self._analysis_lookup = AnalysisLookup(self)
         return self._analysis_lookup
 
+    @property
+    def pydantic_semantic_manifest(self) -> Optional[PydanticSemanticManifest]:
```
Comment (peterallenwebb) [inline core/dbt/contracts/graph/manifest.py:992]: Borderline nitpicky, the return type is actually a non-Optional PydanticSemanticManifest, which is the more attractive return type anyway.

Comment (QMalcolm) [reply core/dbt/contracts/graph/manifest.py:992]: Oh right, good catch! I had planned on having it return `None` if there were not any SemanticModel or Metric nodes 🤦 I think returning an empty semantic model is a pretty good default. I'll get the typing updated, and then also update `write_semantic_manifest` to not have to anticipate it ever being `None`

**Review by peterallenwebb — APPROVED** [review]
Overall this looks good to me as a beta/starting-point. We will be following up the major remaining issues I see in this sprint with:

#7800 
#7821 
#7827

File: core/dbt/contracts/graph/nodes.py [hunk 5]
```diff
@@ -566,7 +568,7 @@ class FileSlice(dbtClassMixin, Replaceable):
 
 
 @dataclass
-class SourceFileMetadata(dbtClassMixin, Replaceable):
+class Metadata(dbtClassMixin, Replaceable):
```
Comment (MichelleArk) [inline core/dbt/contracts/graph/nodes.py:571]: is this rename intentional? re: https://github.com/dbt-labs/dbt-core/pull/7769#discussion_r1222086236

Comment (QMalcolm) [reply core/dbt/contracts/graph/nodes.py:571]: Nope! looks like the merge of `main` did screw some things up 😞
